### PR TITLE
WIP: mixed precision finetuning

### PR DESCRIPTION
**Pull request by benleetownsend**
Key behaviour change here is that we now initialise the GPU on __init__ instead of on .finetune or an inference call.

### CODE REVIEW

File: finetune/util/optimize_loss.py [hunk 1]
```diff
@@ -117,7 +117,7 @@ def optimize_loss(
             )
         )
         tf.compat.v1.summary.scalar(
-            "global_norm/gradient_norm", tf.linalg.global_norm(list(zip(*gradients))[0]),
+            "global_norm/gradient_norm", tf.linalg.global_norm([tf.cast(g[0], tf.float32) for g in gradients]),
```
Comment (benleetownsend) [inline finetune/util/optimize_loss.py]: Gradients should never be fp16... why is this required?

File: finetune/base_models/bert/model.py [hunk 2]
```diff
@@ -185,40 +206,53 @@ class RoBERTa(_BaseBert):
     }
     required_files = [
         {
-            "file": os.path.join(FINETUNE_BASE_FOLDER, "model", "bert", "roberta-model-sm-v2.jl"),
+            "file": os.path.join(
+                FINETUNE_BASE_FOLDER, "model", "bert", "roberta-model-sm-v2.jl"
+            ),
             "url": urljoin(ROBERTA_BASE_URL, "roberta-model-sm-v2.jl"),
```
Comment (chris-wells-13) [inline finetune/base_models/bert/model.py:224]: just a thought, somewhat unrelated to this PR.  should the actual model file names be hardcoded in here?  seems like we'd be better served if they were in a config or, better yet, the config was itself a database so we could modify without making code changes.

File: finetune/datasets/reuters.py [hunk 3]
```diff
@@ -78,9 +80,13 @@ def download(self):
...
         random_state=42
     )
-    model = SequenceLabeler(batch_size=1, n_epochs=3, val_size=0.0, max_length=512, chunk_long_sequences=True, subtoken_predictions=False, crf_sequence_labeling=True, multi_label_sequences=False)
+    model = SequenceLabeler(optimize_for="predict_speed_fp16")##mixed_precision=True, float_16_predict=True, tensorboard_folder="fp16", batch_size=12, predict_batch_size=40)
+    start = time.time()
     model.fit(trainX, trainY)
+#    start = time.time()
     predictions = model.predict(testX)
-    print(predictions)
+    print("Elapsed", time.time() - start)
```
Comment (chris-wells-13) [inline finetune/datasets/reuters.py]: is this a holdover from testing?  if not, should it be run through the logger?  (i guess it is in a test file...probably fine then.)

File: finetune/model.py [hunk 4]
```diff
@@ -101,17 +99,54 @@ def fp16_variable_getter(
...
+            else feat,
+            features,
+        )
+    elif estimator_mode == tf.estimator.ModeKeys.PREDICT and fp16_predict:
         custom_getter = fp16_variable_getter
         features = tf.nest.map_structure(
-            lambda feat: tf.cast(feat, tf.float16) if feat.dtype == tf.float32 else feat,
-            features
+            lambda feat: tf.cast(feat, tf.float16)
+            if feat.dtype == tf.float32
```
Comment (chris-wells-13) [inline finetune/model.py:141]: there is no need to worry about tf.float64 getting out of experimental and sneaking through this is there?

**Review by chris-wells-13 — APPROVED** [review]
this looks good.  some questions, but none of them blocking anything.

File: finetune/target_models/classifier.py [hunk 5]
```diff
@@ -4,7 +4,7 @@
 import tensorflow as tf
 import tensorflow_addons as tfa
 import numpy as np
-from imblearn.over_sampling import RandomOverSampler
+#from imblearn.over_sampling import RandomOverSampler
```
Comment (madisonmay) [inline finetune/target_models/classifier.py]: We should uncomment or consider removing this functionality.

File: finetune/model.py [hunk 6]
```diff
@@ -230,35 +277,43 @@ def _model_fn(features, labels, mode, params):
...
-                        mode=mode
+                        mode=mode,
                     )
                     # No support for any form of text generation for MLM for now
                     lm_predict_op = None
-                else: 
-                    raise FinetuneError("Unsupport `lm_type` option: {}".format(lm_type))
+                else:
+                    raise FinetuneError(
+                        "Unsupport `lm_type` option: {}".format(lm_type)
```
Comment (madisonmay) [inline finetune/model.py]: Totally unrelated to this PR but this should be "unsupported"

File: finetune/model.py (hunk 4 above)
Comment (madisonmay) [reply finetune/model.py:141]: So this is intentionally keeping anything that is fp64 as fp64?  Do we have anything that is / should be fp64?

File: finetune/model.py [hunk 7]
```diff
@@ -101,17 +99,54 @@ def fp16_variable_getter(
     )
 
 
-def get_variable_getter(estimator_mode, features, fp16_predict):
-    if estimator_mode == tf.estimator.ModeKeys.PREDICT and fp16_predict:
+def mp_variable_getter(
```
Comment (madisonmay) [inline finetune/model.py]: Could we name this to something other than mp so my brain stops reading "multiprocessing_variable_getter" haha

File: finetune/model.py (hunk 4 above)
Comment (madisonmay) [reply finetune/model.py:141]: Is it typical to make prediction pure fp16 or is there potential benefit to mixed precision inference?

File: finetune/base_models/bert/model.py (hunk 2 above)
Comment (madisonmay) [reply finetune/base_models/bert/model.py:224]: Since this is open source I don't think a database is the right play.   Maybe an argument to be made that this should be  a JSON or yaml file or something but "code as config" does afford some niceties that yaml / JSON don't.

File: finetune/base.py [hunk 8]
```diff
@@ -92,34 +91,73 @@ def cleanup():
...
+        ):
+            LOGGER.warning(
+                "float_16_predict and mixed_precision only supported by bert based models and Volta or newer GPUs"
+            )
             config.float_16_predict = False
+            config.mixed_precision = False
+
+        if (
+            not issubclass(config.base_model, _BaseBert) or no_fp16
+        ) and "fp16" in config.optimize_for:
```
Comment (madisonmay) [inline finetune/base.py:132]: Interesting -- what's the rationale behind making fp16 be part of the optimize_for string?

File: finetune/base.py [hunk 9]
```diff
@@ -92,34 +91,73 @@ def cleanup():
...
                 raise ValueError("There is no auto setting for {}".format(ak))
             config[ak] = overrides[ak]
+
+        if hasattr(self, "input_pipeline"):
+            self.input_pipeline.config = config
+
+        if hasattr(self, "config"):
+            for k, v in self.config.items():
+                if k not in auto_keys and k != "optimize_for" and config[k] != v:
+                    LOGGER.warning(
```
Comment (madisonmay) [inline finetune/base.py:155]: Wouldn't this end up being quite noisy?  I would expect this to be virtually always hit.

File: finetune/base_models/bert/model.py [hunk 10]
```diff
@@ -99,6 +106,32 @@ def get_optimal_params(cls, config):
...
+        elif config.optimize_for.lower() == "accuracy_fp16":
+            overrides = {
+                "max_length": base_max_length,
+                "n_epochs": base_n_epochs,
+                "batch_size": 12,
+                "chunk_context": None,
+                "predict_batch_size": 40,
+                "mixed_precision": True,
+                "float_16_predict": True,
+                "lr": base_learning_rate * 6,
```
Comment (madisonmay) [inline finetune/base_models/bert/model.py:122]: How is lr rescaling calculated?  Is it something we could auto-apply based on whether or not fp16 is used rather than requiring manual configuration w/ the optimize_for defaults aren't used?  I could see this being a common issue with configuration

File: finetune/base.py (hunk 8 above)
Comment (benleetownsend) [reply finetune/base.py:132]: Primarily just so we don't have to add a another tier of config defaults.

Comment (benleetownsend) [reply finetune/base.py:132]: Resolution is already quite complex. I would agree with potentially renaming this from optimize_for to something like "preset_config" or something but I don't think we should change these automatically when you set mixed_precision = True or anything.

File: finetune/model.py (hunk 4 above)
Comment (benleetownsend) [reply finetune/model.py:141]: If anything is float64 I think we should assume it is intentional and would already be cast before being used in the model. I don't think we have anything that generates fp64 input. Pure float 16 would be slightly preferable over mixed at inference just due to the size of the weights.

File: finetune/base_models/bert/model.py (hunk 10 above)
Comment (benleetownsend) [reply finetune/base_models/bert/model.py:122]: As above. Probably yes, but it would come at added complexity to the config resolution. Personally think if you are departing from the presets there should be as little magic as possible.

File: finetune/base_models/huggingface/models.py [hunk 11]
```diff
@@ -62,10 +62,18 @@
...
         )
     ],
-    config_overrides={"low_memory_mode": True}
+    config_overrides={
+        "low_memory_mode": True,
+        "batch_size": 2,
+        "n_epochs": 16,
+        "lr": 1.68e-3,
+        "max_grad_norm": 1.1,
+        "lr_warmup": 0.7,  # Not a typo - just weird.
```
Comment (madisonmay) [inline finetune/base_models/huggingface/models.py:74]: Someone used bayesian hyperparam opt :P

File: finetune/base_models/bert/model.py (hunk 10 above)
Comment (madisonmay) [reply finetune/base_models/bert/model.py:122]: I'm happy with that decision.

**Review by madisonmay — APPROVED** [review]
